sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+
+    // Test writing attributes when others are already enabled.
+    ExpectAttrFieldUpdate(attr_field, initial_attrs, new_attrs,
+                          kAttrFieldUpdateWrite);
+
+    EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, pad_kind,
+                                      attr_field.pad_num, new_attrs),
+              kDifPadctrlChangeOk);
+
+    // For every attribute test writing it when other attributes are already
+    // enabled
+    for (auto attr : attributes_) {
+      ExpectAttrFieldUpdate(attr_field, initial_attrs, attr,
+                            kAttrFieldUpdateWrite);
+
+      EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, pad_kind,
+                                        attr_field.pad_num, attr),
+                kDifPadctrlChangeOk);
+    }
+  }
+
+  /**
+   * Tests a failure to enable an attribute due to a conflict
+   *
+   * @param attr_field The attribute field to test
+   * @param pad_kind What kind of pad `attr_field` is for
+   */
+  void TestEnableAttrFailureConflict(const AttrField &attr_field,
+                                     dif_padctrl_pad_kind_t pad_kind) {
+    DeviceInit();
+
+    ExpectLockCheck(false);
+
+    // Expect a read of `attr_field` with a couple of attributes set.
+    ExpectAttrFieldRead(attr_field, dif_padctrl_attr_t(kDifPadctrlAttrIOInvert |
+                                                       kDifPadctrlAttrPullUp));
+
+    // Attempt to set an attribute that conflicts with the currently set ones
+    // expecting the appropriate error.
+    EXPECT_EQ(
+        dif_padctrl_enable_attrs(&dif_padctrl_, pad_kind, attr_field.pad_num,
+                                 kDifPadctrlAttrPullDown),
+        kDifPadctrlChangeAttrConflict);
+  }
+
+  /**
+   * Tests a failure to write attributes due to a conflict
+   *
+   * @param attr_field The attribute field to test
+   * @param pad_kind What kind of pad `attr_field` is for
+   */
+  void TestWriteAttrFailureConflict(const AttrField &attr_field,
+                                    dif_padctrl_pad_kind_t pad_kind) {
+    DeviceInit();
+
+    ExpectLockCheck(false);
+
+    // Expect a read of `attr_field` with an attribute set.
+    ExpectAttrFieldRead(attr_field,
+                        dif_padctrl_attr_t(kDifPadctrlAttrIOInvert));
+
+    // Write back attributes that conflict with each other and not the attribute
+    // that was initially set expecting the appropriate error.
+    dif_padctrl_attr_t conflicting_attrs =
+        dif_padctrl_attr_t(kDifPadctrlAttrPullUp | kDifPadctrlAttrPullDown);
+
+    EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, pad_kind,
+                                      attr_field.pad_num, conflicting_attrs),
+              kDifPadctrlChangeAttrConflict);
+  }
+
+  dif_padctrl_t dif_padctrl_;
+};
+
+TEST_F(DifPadctrlTest, InitSuccess) { DeviceInit(); }
+
+TEST_F(DifPadctrlTest, InitFailureInvalidPadctrl) {
+  mmio_region_t base_addr = dev().region();
+
```

Comment:
I think that the rest of this file is a bit too heavy-handed with the line-breaks. Consider only using line breaks to break up unrelated stanzas of code where otherwise you would have quite a few expressions in a row that are unrelated. For example, I would have written this code as
```
mmio_region_t base_addr = ...;
EXPECT_EQ(...);
```
(In this particular case, though, I would inline `base_addr`, since having it out-of-line like that doesn't add much in the way of extra context.)

Elsewhere, I would avoid keeping `DeviceInit();` on its own stanza, or keeping `EXPECT_EQ(...)` calls separate from your `Expect` helpers.

Reply:
I have done various removals of vertical whitespace